rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+---
+feature: system-parameters
+authors:
+  - "Eric Fu"
+start_date: "2023/01/11"
+---
+
+# RFC: System Parameters
+
+## Motivation
+
+Currently, we have 3 kinds of configurations:
+
+- Command-line arguments
+- Configuration file `risingwave.toml`
+- Session parameters
+
+These can not meet the requirements for several reasons.
+
+1. **Consistency**. `risingwave.toml` is a local file, which makes it possible that multiple nodes could see different parameters and this is absolutely unacceptable. For example, to configure the remote directory for storing data, users may use the `data_directory` parameter. If some nodes use `data_directory = 'hummock_001'` and others use `data_directory = 'hummock_002'`, the stored data will be corrupted.
+2. **Mutability**. Users are unable to `SET` a config after the process started. The root cause is still the local file - changes of configs must be persisted in some cluster-level storage.
+
+This RFC proposes to introduce centralized configurations on Meta Service and data will be persisted in etcd.
+
+For the sake of clarity, in this doc, we will use the term **"parameters"** for system and session parameters, and **"config"** for command-line arguments or config files.
+
+**Note:** Previously I think this might not be an urgent issue until I realized that our Cloud platform couples tightly with more and more configurations. The later we commence, the heavier burden it will become. 
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Parameters (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Parameters
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
+
+Here we could borrow the syntax from Postgres:
+
+```sql
+-- Set or reset a parameter
+ALTER SYSTEM / DATABASE SET parameter TO value / 'value' / DEFAULT;
+
+-- List all parameters keys & values
+SHOW PARAMETERS;
+```
+
+> **Note**: PG uses `SHOW ALL` because it doesn't has other `SHOW` commands like us. Which one is better?
+
+
+References:
+
+- [PostgreSQL: Documentation: 15: ALTER SYSTEM](https://www.postgresql.org/docs/current/sql-altersystem.html)
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+- [PostgreSQL: Documentation: 15: SHOW](https://www.postgresql.org/docs/current/sql-show.html)
+
+Similar to other system metadata e.g. catalog,
+
+- Changes to global parameters should be propagated to all nodes, in the cluster in a deterministic order ideally
```

Comment:
Considering the nature of distributed systems, there will be cases where multiple values of the same parameter exist on various nodes, no matter how we optimize the notification mechanism. Will we introduce things like what is described in the Online Schema Change paper to ensure consistency and atomicity?

Reply:
Do you mean [Online, Asynchronous Schema Change in F1](https://storage.googleapis.com/pub-tools-public-publication-data/pdf/41376.pdf)  by Rae et al.? As I understand [this summary](https://pages.cs.wisc.edu/~zuyu/summaries/cs764/f1Schema) they basically use 2PC to change schemas, right? 

Maybe we could still rely on ETCD (with or without abstracting via a K8s CM), since it offers strong consistency? A 2PC using ETCD could look like this maybe: 

1. Client issues settings change against system 
2. ETCD changes configs 
3. Pods are alerted about config change (if we use CM, K8s would do that for us)
4. Meta (or some other components) asks each pod if it is ready to apply the changes. If all reply yes, then we apply. 

Just a suggestion :)

Reply:
Yep, it's the exact paper that I referred.

>Maybe we coul rely on ETCD (with or without abstracting via a K8s CM), since it offers strong consistency?

🤔 Sounds like a feasible solution.